### ISSUE TITLE
Traefik 1 labels seem to prevent Website editing

### ISSUE THREAD

**Issue by gdgellatly** [Tecnativa/doodba-copier-template#166]
So I was getting all sorts of errors trying to click edit on the frontend website in prod but not dev. Tried a few things, including a copier update -D.

Eventually just deleted all the Traefik labels I thought could cause it as well as all the Traefik 2 labels and it worked. Then I noticed there was 301 redirect happening for a POST to `/web/dataset/call_kw` to load the Wysiwyg Editor assets. Which was creating a typical endpotini error about calling with http instead of JSON.

Working back through the labels. Eventually we found this one causes the issue, and commenting out it works, although actually I think it might not be the real culprit, although I don't know. For avoidance of doubt <your host> is anonymisation not the actual label.

```yaml
traefik.forbiddenCrawlers-0.frontend.rule: Host:<your host>;Path:/web{anything:/.*},/website/info{anything:/.*}
```

EDIT: Sorry missing details Traefik 1.7, 13.0 Doodba. Pretty Vanilla setup in terms of extra mods and the like, all recently updated.

**Comment by Yajo** [comment]
Please provide all the labels, and traefik logs in debug mode. You can anonymize the domains if you want.

**Comment by gdgellatly** [comment]
Hi,

Traefik generates no logs in INFO level, in DEBUG, on a production server with hundreds of users it is basically impossible to do a spot check in business hours. 

I'll give it a go later tonight.

It generates 50,000 log entries a minute, I can see the crashmanager entries but nothing notable for anything else.

The Traefik labels are: These *should* be the stock standard after deleting them all and regenerating.

```yaml
      doodba.domain.main: "o4sb.com"
      traefik.enable: "true"
      traefik.domain: "o4sb.com"
      traefik.forbiddenCrawlers-0.frontend.headers.customResponseHeaders:
        "X-Robots-Tag:noindex, nofollow"
      traefik.forbiddenCrawlers-0.frontend.rule: Host:o4sb.com;Path:/web{anything:/.*},/website/info{anything:/.*}
      traefik.main-0.frontend.rule: Host:o4sb.com
      traefik.longpolling-0.frontend.rule: Host:o4sb.com;Path:/longpolling/{anything:.*}
      traefik.longpolling-0.port: 8072
      traefik.alt-1.frontend.redirect.regex: ^(.*)://([^/]+)/(.*)$$
      traefik.alt-1.frontend.redirect.replacement: $$1://o4sb.com/$$3
      traefik.alt-1.frontend.rule: Host:odoo.nz, www.odoo.nz, www.o4sb.com
      traefik.http.middlewares.o4sb-odoo-13-0-prod-buffering.buffering.retryExpression:
        IsNetworkError() && Attempts() < 5
      traefik.http.middlewares.o4sb-odoo-13-0-prod-compress.compress: "true"
      ? traefik.http.middlewares.o4sb-odoo-13-0-prod-forbid-crawlers.headers.customResponseHeaders.X-Robots-Tag
      : "noindex, nofollow"
      traefik.http.middlewares.o4sb-odoo-13-0-prod-forceSecure.headers.forceSTSHeader: "true"
      traefik.http.middlewares.o4sb-odoo-13-0-prod-forceSecure.headers.sslRedirect: "true"
      traefik.http.services.o4sb-odoo-13-0-prod-main.loadbalancer.server.port: 8069
      traefik.http.services.o4sb-odoo-13-0-prod-longpolling.loadbalancer.server.port: 8072
      traefik.http.routers.o4sb-odoo-13-0-prod-main-0.rule: Host(`o4sb.com`)
      traefik.http.routers.o4sb-odoo-13-0-prod-main-0.service: o4sb-odoo-13-0-prod-main
      traefik.http.routers.o4sb-odoo-13-0-prod-main-0.middlewares:
        o4sb-odoo-13-0-prod-buffering, o4sb-odoo-13-0-prod-compress,
        o4sb-odoo-13-0-prod-forceSecure
      traefik.http.routers.o4sb-odoo-13-0-prod-main-secure-0.rule: Host(`o4sb.com`)
      traefik.http.routers.o4sb-odoo-13-0-prod-main-secure-0.service: o4sb-odoo-13-0-prod-main
      traefik.http.routers.o4sb-odoo-13-0-prod-main-secure-0.middlewares:
        o4sb-odoo-13-0-prod-buffering, o4sb-odoo-13-0-prod-compress,
        o4sb-odoo-13-0-prod-forceSecure
      traefik.http.routers.o4sb-odoo-13-0-prod-main-secure-0.tls: "true"
      traefik.http.routers.o4sb-odoo-13-0-prod-main-secure-0.tls.certResolver: letsencrypt
      traefik.http.routers.o4sb-odoo-13-0-prod-longpolling-0.rule:
        Host(`o4sb.com`) && PathPrefix(`/longpolling/`)
      traefik.http.routers.o4sb-odoo-13-0-prod-longpolling-0.service: o4sb-odoo-13-0-prod-longpolling
      traefik.http.routers.o4sb-odoo-13-0-prod-longpolling-0.middlewares: o4sb-odoo-13-0-prod-forceSecure
      traefik.http.routers.o4sb-odoo-13-0-prod-longpolling-secure-0.rule:
        Host(`o4sb.com`) && PathPrefix(`/longpolling/`)
      traefik.http.routers.o4sb-odoo-13-0-prod-longpolling-secure-0.service: o4sb-odoo-13-0-prod-longpolling
      traefik.http.routers.o4sb-odoo-13-0-prod-longpolling-secure-0.middlewares: o4sb-odoo-13-0-prod-forceSecure
      traefik.http.routers.o4sb-odoo-13-0-prod-longpolling-secure-0.tls: "true"
      traefik.http.routers.o4sb-odoo-13-0-prod-longpolling-secure-0.tls.certResolver: letsencrypt
      traefik.http.routers.o4sb-odoo-13-0-prod-forbiddenCrawlers-0.rule:
        Host(`o4sb.com`) && (PathPrefix(`/web/`, `/website/info/`) || Path(`/web`,
        `/website/info`))
      traefik.http.routers.o4sb-odoo-13-0-prod-forbiddenCrawlers-0.service: o4sb-odoo-13-0-prod-main
      traefik.http.routers.o4sb-odoo-13-0-prod-forbiddenCrawlers-0.middlewares:
        o4sb-odoo-13-0-prod-buffering, o4sb-odoo-13-0-prod-compress,
        o4sb-odoo-13-0-prod-forbid-crawlers, o4sb-odoo-13-0-prod-forceSecure
      traefik.http.routers.o4sb-odoo-13-0-prod-forbiddenCrawlers-secure-0.rule:
        Host(`o4sb.com`) && (PathPrefix(`/web/`, `/website/info/`) || Path(`/web`,
        `/website/info`))
      traefik.http.routers.o4sb-odoo-13-0-prod-forbiddenCrawlers-secure-0.service: o4sb-odoo-13-0-prod-main
      traefik.http.routers.o4sb-odoo-13-0-prod-forbiddenCrawlers-secure-0.middlewares:
        o4sb-odoo-13-0-prod-buffering, o4sb-odoo-13-0-prod-compress,
        o4sb-odoo-13-0-prod-forbid-crawlers, o4sb-odoo-13-0-prod-forceSecure
      traefik.http.routers.o4sb-odoo-13-0-prod-forbiddenCrawlers-secure-0.tls: "true"
      traefik.http.routers.o4sb-odoo-13-0-prod-forbiddenCrawlers-secure-0.tls.certResolver: letsencrypt
      traefik.http.middlewares.o4sb-odoo-13-0-prod-redirect-1.redirectRegex.regex: ^(.*)://([^/]+)/(.*)$$
      traefik.http.middlewares.o4sb-odoo-13-0-prod-redirect-1.redirectRegex.replacement: $$1://o4sb.com/$$3
      traefik.http.routers.o4sb-odoo-13-0-prod-redirect-1.rule:
        Host(`odoo.nz`, `www.odoo.nz`, `www.o4sb.com`)
      traefik.http.routers.o4sb-odoo-13-0-prod-redirect-1.service: o4sb-odoo-13-0-prod-main
      traefik.http.routers.o4sb-odoo-13-0-prod-redirect-1.middlewares:
        o4sb-odoo-13-0-prod-compress, o4sb-odoo-13-0-prod-forceSecure,
        o4sb-odoo-13-0-prod-redirect-1
      traefik.http.routers.o4sb-odoo-13-0-prod-redirect-secure-1.rule:
        Host(`odoo.nz`, `www.odoo.nz`, `www.o4sb.com`)
      traefik.http.routers.o4sb-odoo-13-0-prod-redirect-secure-1.service: o4sb-odoo-13-0-prod-main
      traefik.http.routers.o4sb-odoo-13-0-prod-redirect-secure-1.middlewares:
        o4sb-odoo-13-0-prod-compress, o4sb-odoo-13-0-prod-forceSecure,
        o4sb-odoo-13-0-prod-redirect-1
      traefik.http.routers.o4sb-odoo-13-0-prod-redirect-secure-1.tls: "true"
      traefik.http.routers.o4sb-odoo-13-0-prod-redirect-secure-1.tls.certResolver: letsencrypt

```
From browser with crawler traefik label enabled:  Output it calls render_template on ir.ui.view for website.compiled_assets_wysiwyg. Standard jsonrpc call.
```
Request URL: https://o4sb.com/web/dataset/call_kw/
Request Method: POST
Status Code: 301 
Remote Address: 172.16.10.4:443
Referrer Policy: strict-origin-when-cross-origin
Response Headers:
content-length: 0
date: Tue, 10 Nov 2020 21:50:36 GMT
location: /web/dataset/call_kw
status: 301
```

Gets redirected to a GET call over http.

Same call with just the forbiddenCrawlers-0.frontend.rule disabled
```
content-length: 503
content-type: application/json
date: Tue, 10 Nov 2020 22:01:50 GMT
server: Werkzeug/0.14.1 Python/3.6.12
set-cookie: session_id=540b79a93f6becfa188288842287e44421335776; Expires=Mon, 08-Feb-2021 22:01:50 GMT; Max-Age=7776000; HttpOnly; Path=/
status: 200
```

Traefik Config is pretty vanilla
```yaml
services:
  proxy:
    image: traefik:1.7-alpine
    networks:
      shared:
      private:
      public:
    volumes:
      - acme:/etc/traefik/acme:rw,Z
    ports:
      - "80:80"
      - "443:443"
    depends_on:
      - dockersocket
    restart: unless-stopped
    privileged: true
    tty: true
    command:
      - --ACME.ACMELogging
      - --ACME.Email=<my email>
      - --ACME.EntryPoint=https
      - --ACME.HTTPChallenge.entryPoint=http
      - --ACME.OnHostRule
      - --ACME.Storage=/etc/traefik/acme/acme.json
      - --DefaultEntryPoints=http,https
      - --EntryPoints=Name:http Address::80 Redirect.EntryPoint:https
      - --EntryPoints=Name:https Address::443 TLS
      - --LogLevel=INFO
      - --Docker
      - --Docker.EndPoint=http://dockersocket:2375
      - --Docker.ExposedByDefault=false
      - --Docker.Watch
```

**Comment by gdgellatly** [comment]
Interestingly, I can also reproduce similar on traefik 1.7 on a different staging server, with similar outcomes.

Vanilla config EXCEPT copier wasn't adding the main frontend rule so added manually.

With forbid-crawlers rule - too many redirects
Without - works as expected.

**Comment by carlosecv** [comment]
This is what im trying to figure it out the traefik labels are not the same as we had on previous copier version, this affects on widget="signature" field on views

**Comment by gdgellatly** [comment]
Yes, email marketing module as well in backend.

**Comment by Yajo** [comment]
Please try these 2 solution, and let's see if any of them work for you:

Solution 1:

```diff
-      traefik.forbiddenCrawlers-0.frontend.rule: Host:o4sb.com;Path:/web{anything:/.*},/website/info{anything:/.*}
+      traefik.forbiddenCrawlers-0.frontend.rule: Host:o4sb.com;Path:/web{anything:(/|/.*)},/website/info{anything:(/|/.*)}
```

Solution 2:

```diff
-      traefik.forbiddenCrawlers-0.frontend.rule: Host:o4sb.com;Path:/web{anything:/.*},/website/info{anything:/.*}
+      traefik.forbiddenCrawlers-0.frontend.rule: Host:o4sb.com;Path:/web,/website/info;PathPrefix:/web/,/website/info/
```

Next, also something quite wierd I see is that your RPC call is being done to `https://o4sb.com/web/dataset/call_kw/` and not to `https://o4sb.com/web/dataset/call_kw`. AFAIK Odoo automatically will strip trailing slashes always. Are you sure it's not Odoo who is issuing the redirection?

**Comment by gdgellatly** [comment]
@Yajo 

Firstly thanks for taking the time to look at this. There is nothing especially special about this site, and in any case, I can replicate it on vanilla and multiple other instances. The only thing I can think of is `oca_web_pwa`, although if I unregister the service worker I get the same issue. Also that module is not installed or even available on all instances. 

Also in all cases if I disable that rule, it just works. Thats what points to me this rule being the issue.

As for the trailing slash, that is a specifically formed URL in Odoo's web client, it gets created and passed via a js ajax call, so I guess the always part of the statement does not hold true for posts to JSON endpoints, although maybe it strips server side.

Anyway - 
Solution 1: Fails both with and without PWA and also in plain vanilla instance.
Solution 2: Works in all cases.

**Comment by gdgellatly** [comment]
More info, just had a client call regards their staging server with same issue in test.yaml. Equivalent modification to test.yaml of same rule was required. (Basically removing everything in braces.

**Comment by Yajo** [comment]
In #189 I've been trying to fix, but it turns out that solution 2 is mostly disabling the rule (as it will never do a match). Please try solution 3 and report back:

```diff
-      traefik.forbiddenCrawlers-0.frontend.rule: Host:o4sb.com;Path:/web{anything:/.*},/website/info{anything:/.*}
+      traefik.forbiddenCrawlers-0.frontend.rule: Host:o4sb.com;Path:/web,/web/{anything:.*},/website/info,/website/info/{anything:.*}
```

Thanks!

**Comment by gdgellatly** [comment]
@Yajo It doesn't work. 

So what we know is that it occurs on a naked call to call_kw when there is a post with a trailing slash. This occurs in v13 during loadAsset of ajax.js where it renders a template.

Specifically is this code here

```javascript
var loadAsset = (function () {
    var cache = {};

    var load = function loadAsset(xmlId, context, tplRoute = '/web/dataset/call_kw/') {
        if (cache[xmlId]) {
            return cache[xmlId];
        }
        context = _.extend({}, odoo.session_info.user_context, context);
        const params = {
            args: [xmlId, {
                debug: config.isDebug()
            }],
            kwargs: {
                context: context,
            },
        };
        if (tplRoute === '/web/dataset/call_kw/') {
            Object.assign(params, {
                model: 'ir.ui.view',
                method: 'render_template',
            });
        }
```

The way I see it there is potentially some solutions.
1 - It is an Odoo V13 bug and needs fixing upstream?
2 - We exclude calls to just /web/dataset/call_kw/ from matching in the rule somehow (maybe `.+` instead of `.*`)
3- Publish Traefik 2 config and deprecate 1.

**Comment by Yajo** [comment]
So, what are the *specific* steps I have to execute through UI to trigger this failure? I mean: install module X, go here, click there...

**Comment by gdgellatly** [comment]
Hi @Yajo 

On v13
Install Website.
Login.
Go to Website
Click Edit.

Installing mass_mailing and creating a Mass Mailing is another way.

**Comment by Yajo** [comment]
I'm sorry, I've done that and I cannot reproduce it :shrug: 

All requests to `call_kw` return a 200 OK, with proper body, behind a Traefik 1.7, both for test and prod envs:

![imagen](https://user-images.githubusercontent.com/973709/100590329-25a25b80-32ec-11eb-95f2-f14b3f0a32c1.png)

UI works flawlessly. You might be having some other kind of error. Please check the labels reported by `docker-compose config`, which is the real config used after combining the several definitions.

My rules in `prod.yaml` are like yours, except for the obvious different domain.

This seems some error specific to your deployment. Another possibility if you can't debug any further is to just update to Traefik 2, as it is supported by the template and works better overall.

I'll close because it doesn't seem related to the template itself but to some downstream problem.

**Comment by gdgellatly** [comment]
@yajo I am sure the issue isn't downstream. It happens on different servers, different versions, no modules installed, it happens even on a freshly blown away config. You even have multiple reports from different people.

That said, I am happy to use Traefik 2, however you have not documented the config to use with your labels, which I think is why loads have stuck with v1.x for doodba..

**Comment by Yajo** [comment]
Well, as long as I can't reproduce, I can't consider it a bug in our side, I hope you understand. :shrug: 

Maybe you can just create a subproject with everything, putting `127.0.0.1.sslip.io` as domain, so I can boot a traefik server locally and let it route the requests locally, and publish it as a gist, so I can just clone it, `docker-compose up -d` and see how it fails. :thinking: I really wish I could reproduce it because as you said we have several reports, but none of them is helpful enough ATM. :confused: 

Traefik 2 allows for much deeper and flexible configuration than 1, so that's why I left the operator to decide how to deploy it. Also it has the problem of global per-provider namespaces (https://github.com/traefik/traefik/issues/5134) instead of per-container as Traefik 1, so it's not just a matter of switching the flip from 1 to 2... It requires testing from your operations team.

However I'd be glad to accept some PR on the docs if you feel this part is confusing and could be improved. :blush:

**Comment by carlosecv** [comment]
Confirmed:

      #traefik.forbiddenCrawlers-0.frontend.headers.customResponseHeaders:
      #  "X-Robots-Tag:noindex, nofollow"
      #traefik.forbiddenCrawlers-0.frontend.rule: Host:mydomain.com.mx;Path:/web{anything:(/|/.*)},/website/info{anything:(/|/.*)}

Commenting this lines we can edit website again

This confirms that we need to improve labels creation and get more documented about this.

**Comment by sotolucas** [comment]
@Yajo same issue here, with a fresh install.
